src/net_processing.cpp

```diff
@@ -5639,7 +5633,7 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                             if (!tx_relay->m_bloom_filter->IsRelevantAndUpdate(*txinfo.tx)) continue;
                         }
                         tx_relay->m_tx_inventory_known_filter.insert(hash);
-                        // Responses to MEMPOOL requests bypass the m_recently_announced_invs filter.
+                        tx_relay->m_recently_announced_invs.insert(hash);
```

Comment:
Doesn't this just change the fingerprinting vector from `GETDATA txid` to `FILTERCLEAR; FILTERADD whatever; MEMPOOL` ? In which case we're just as easily fingerprinted, but we spend more cpu while being fingerprinted?

Reply:
Ah, I see what you mean. That makes sense.

Next point on those lines though: adding many things to `m_recently_announced_invs` can overflow the bloom filter, either causing false positives (leading to privacy bugs), or since it's a rolling bloom filter, causing overflow leading to false negatives (leading to functionality bugs where you can't get the txs from the MEMPOOL response). Wouldn't it be better to make the insertion conditional on `txinfo.m_time <= now - UNCONDITIONAL_RELAY_DELAY` ? Even then this approach seems like it can trigger false positives fairly easily, since the rolling bloom filter capacity is only 3500 txs?

Reply:
> Wouldn't it be better to make the insertion conditional on `txinfo.m_time <= now - UNCONDITIONAL_RELAY_DELAY` ? 

Also wouldn't that also apply here? https://github.com/bitcoin/bitcoin/blob/27c40e3dcf280a137bcfaa787674c26781d911b8/src/net_processing.cpp#L5688

Reply:
`MAX_PEER_TX_ANNOUNCEMENTS` doesn't really provide a rate limit -- it stops the queue from being more than 5k at any point in time, but the flow through the queue could still be 400tx/s, for instance.  Beyond that, `m_recently_announced_invs` only tracks things we announced, not things we received, and we rate limit that (as per the static_assert on `INVENTORY_MAX_RECENT_RELAY`).

Delaying MEMPOOL results for 2 minutes seems annoying. But how about just recording the timestamp of the last MEMPOOL request into the `txrelay` structure, and skipping the bloom lookup if `m_time <= max(now-2min, txrelay->last_mempool_request)` ?

You could make the inv-trickle insertion into the bloom filter conditional too, but I think that would so rarely actually have an effect that it wouldn't matter?

Reply:
> But how about just recording the timestamp of the last MEMPOOL request ...

Patch might look something like...

<details><summary>Details</summary>
```diff
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -295,6 +295,8 @@ struct Peer {
          *  permitted if the peer has NetPermissionFlags::Mempool or we advertise
          *  NODE_BLOOM. See BIP35. */
         bool m_send_mempool GUARDED_BY(m_tx_inventory_mutex){false};
+        std::chrono::microseconds m_last_mempool_time GUARDED_BY(m_tx_inventory_mutex){0};
+
         /** The next time after which we will send an `inv` message containing
          *  transaction announcements to this peer. */
         std::chrono::microseconds m_next_inv_send_time GUARDED_BY(m_tx_inventory_mutex){0};
@@ -2258,7 +2260,7 @@ CTransactionRef PeerManagerImpl::FindTxForGetData(const Peer::TxRelay& tx_relay,
     auto txinfo = m_mempool.info(gtxid);
     if (txinfo.tx) {
         // If a TX is older than UNCONDITIONAL_RELAY_DELAY, permit the request unconditionally.
-        if (txinfo.m_time <= now - UNCONDITIONAL_RELAY_DELAY) {
+        if (txinfo.m_time <= now - UNCONDITIONAL_RELAY_DELAY || WITH_LOCK(tx_relay.m_tx_inventory_mutex, return txinfo.m_time <= tx_relay.m_last_mempool_time)) {
             return std::move(txinfo.tx);
         }
     }
@@ -5619,6 +5621,7 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                     tx_relay->m_send_mempool = false;
                     const CFeeRate filterrate{tx_relay->m_fee_filter_received.load()};
 
+                    tx_relay->m_last_mempool_time = current_time;
                     LOCK(tx_relay->m_bloom_filter_mutex);
 
                     for (const auto& txinfo : vtxinfo) {
@@ -5632,7 +5635,6 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                         if (tx_relay->m_bloom_filter) {
                             if (!tx_relay->m_bloom_filter->IsRelevantAndUpdate(*txinfo.tx)) continue;
                         }
-                        tx_relay->m_tx_inventory_known_filter.insert(hash);
                         tx_relay->m_recently_announced_invs.insert(hash);
                         vInv.push_back(inv);
                         if (vInv.size() == MAX_INV_SZ) {
```
</details>

Reply:
Okay, other idea: how about just making sure you only send at most 3500 mempool entries younger than 2 minutes, in response to a MEMPOOL message, and add them all to the filter, and don't worry if that causes overflow on results from a previous response?

Reply:
> I wonder if we can remove mapRelay and as a result can get more efficient data structures to remember (w)txids per Peer, knowing that only transaction in the mempool can be in them.

`mapRelay` is mostly what's in the mempool, but can also contain entries that were announced less than 15min ago + fell out of the mempool since then. This is sometimes good for fingerprinting :shrug: but also seems like largely redundant information to me.

Reply:
Knowing when a tx is removed from the mempool means you know exactly when the tx that replaced it was added to the mempool... Txs are also removed from the mempool when a block comes in; if we process the block first, then see a GETDATA for a tx in the block, replying with the tx is better for block propagation (including compact blocks) that sending a NOTFOUND...

Reply:
I don't think the timing of removal from `mapRelay` reveals a secret -- that's 15minutes after it was added, but the time it was added is the time it was first announced (probably to an outbound peer), which is already randomised/batched, compared to the time we received it?

Reply:
What I mean is
Leak available without mapRelay:
1. spy sends txA. It's added to mempool.
2. spy sends txB (through the network) and continuously queries our node for txA
3. once we accept txB to mempool, we serve notfound for txA (because it's not there anymore). spy learns exactly what time we received txB.

^This is prevented with mapRelay because after txA leaves mempool, we'll still be able to serve it from mapRelay.

This is still possible with mapRelay:
1. spy sends txA. It's added to mempool + mapRelay.
2. 15 minutes later, txA expires from mapRelay
3. spy sends txB (through the network) and continuously queries our node for txA
4. once we accept txB to mempool, we serve notfound for txA (because it's not in mempool or mapRelay anymore). spy learns exactly what time we received txB.

Reply:
I think those are rate limited so they don't overflow in less than two minutes, after which they are not relevant anymore.

For outbound connections, INVs are sent out every 2s (`OUTBOUND_INV_BROADCAST_INTERVAL`), while for inbound, they are sent every 5s (`OUTBOUND_INV_BROADCAST_INTERVAL`). In a two-minute timespan, that results in 2100 and 840 INVs at max<sup>1</sup>, respectively. The `m_recently_announced_invs` rolling filter is only checked in `FindTxForGetData` after making sure the traction cannot be sent unconditionally (i.e. after 2min), so if the filter overflows after that it's not an issue.

<sup>1</sup> data is also added to `m_recently_announced_invs` when relaying a transaction with a recent but unconfirmed parent (https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L2343). I'm unsure if that is rate-limited at all, or if the remaining 1400 slots (in the worse case) are to account for that not overflowing.

Reply:
The scenario here is:

 * issue MEMPOOL with a large  , returning 6000 txs from the mempool, all of which are added to the bloom filter, which then causes the first ~750 or more to be removed from the rolling bloom filter as it rolls over
 * request one of those txs via GETDATA, which then fails if the tx hasn't been in the mempool for more than 2 minutes

In normal relay, over the course of two minutes we'd only advertise ~840 txs (~2100 for an outbound), and only add the txs to the bloom filter when advertised, so the rollover would only happen very rarely (less than 1-in-a-million times for an outbound, fewer for an inbound). After #27610 we'll relay more txs during spikes, so these error rates can increase, though.

Reply:
> Side note: The 50k max inv size also seems to overflow, which is probably another source for false negatives, depending on the use case.

You lost me there. Do you mean this may be a possible source of false negatives for the current path, or for both?

In this patch, I can see that potentially being an issue (that's what the original concern from AJ's comes from), but I don't see how that is an issue for master, given nothing is added to `m_recently_announced_invs` when building this huge INVs as a result of a MEMPOOL message.

Reply:
All good, I was just wondering if I may have been missing something.